### PR TITLE
Feature/attri2vec

### PR DESCRIPTION
**Pull request by daokunzhang**
In this branch, I implemented the attri2vec model and evaluate its performance in inferring the representations of out-of-sample nodes through the out-of-sample link prediction task.

I made following three changes:

1. I added the stellargraph/demos/embeddings/stellargraph-attri2vec-DBLP.ipynb file, which shows the work flow for training attri2vec on in-sample subgraphs, and inferring the representations of out-of-sample nodes, following by the out-of-sample link prediction experiments.

2. I added the attri2vecLinkGenerator class to the stellargraph/stellargraph/mapper/link_mappers.py, which is used to generate batches of (feature of target node, id of context node, label of node pair) lists for training the attri2vec model.

3. I changed the stellargraph/demos/embeddings/README.md file, by adding the description for the attri2vec demo.

### CODE REVIEW

**Comment by PantelisElinas** [comment]
Hi @daokunzhang 

please update the code to wrap attri2vec into a layer, write unit tests, and update documentation so we can review.

Regards,

P.

**Comment by daokunzhang** [comment]
Hi @PantelisElinas ,

I have made following updates for the code:

1. Write a Attri2Vec class in file stellargraph/stellargraph/layer/attri2vec.py to wrap attri2vec in a layer, and update the stellargraph/stellargraph/layer/__init__.py file to help import Attri2Vec from the package stellargraph.layer.

2. Write a Attri2VecNodeGenerator class at file stellargraph/stellargraph/mapper/node_mappers.py to predict node representations with the learned attri2vec model.

3. Rewrite the demo file --stellargraph/demos/embeddings/stellargraph-attri2vec-DBLP.ipynb-- with the updated layers and mappers.

4. Write the unit test for Attri2Vec class at file stellargraph/tests/layer/test_attri2vec.py, the unit test for Attri2VecLinkGenerator at file stellargraph/tests/test_linkmappers.py, the unit test for Attri2VecNodeGenerator at file stellargraph/tests/test_nodemappers.py

5. Update the stellargraph/docs/api.txt file to add the index to the new added class: Attri2Vec, Attri2VecLinkGenerator, Attri2VecNodeGenerator.

Best regards!
Daokun

File: stellargraph/layer/attri2vec.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        node_num=None,
+        bias=False,
+        dropout=0.0,
+        normalize=None,
+    ):
+
+        if normalize == "l2":
+            self._normalization = Lambda(lambda x: K.l2_normalize(x, axis=-1))
+
+        elif normalize is None or normalize == "none" or normalize == "None":
```
[the remark this answers is not in the document]
Comment (daokunzhang) [reply stellargraph/layer/attri2vec.py]: For the normalize option check, I follow the code of GraphSAGE layer, to make the implementation of Attri2Vec layer consistent with GraphSAGE.

File: stellargraph/layer/attri2vec.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+                )
+
+            self.input_feature_size = feature_sizes.popitem()[1]
+            self.input_node_num = generator.generator.graph.number_of_nodes()
+
+        elif input_dim is not None and node_num is not None:
+            self.input_feature_size = input_dim
+            self.input_node_num = node_num
+
+        else:
```
[the remark this answers is not in the document]
Comment (daokunzhang) [reply stellargraph/layer/attri2vec.py]: The Attri2Vec layer can be constructed from a generator, which only passes the parameter of the number of nodes and the dimension of features. The Attri2Vec layer can also be directly constructed with these two parameters. To train the Attri2Vec model, a generator is still needed to be passed to the keras.model.fit_generator() function.

File: stellargraph/layer/attri2vec.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+            raise RuntimeError(
+                "If generator is not provided, input_dim and node_num must be specified."
+            )
+
+        # Model parameters
+        self.n_layers = len(layer_sizes)
+        self.bias = bias
+        self.dropout = dropout
+        self.activation = "sigmoid"  # the activation fuction used in hidden layers
```
[the remark this answers is not in the document]
Comment (daokunzhang) [reply stellargraph/layer/attri2vec.py]: It can be other options, like ReLU. We can make it a parameter for init to be specified.

File: stellargraph/mapper/node_mappers.py [hunk 4]
```diff
@@ -508,6 +514,124 @@ def flow_from_dataframe(self, node_targets, shuffle=False):
         )
 
 
+class Attri2VecNodeGenerator:
```
[the remark this answers is not in the document]
Comment (daokunzhang) [reply stellargraph/mapper/node_mappers.py]: To make the name more informative, shall we name it as FeatureBasedNodeGenerator?

File: stellargraph/mapper/node_mappers.py [hunk 5]
```diff
@@ -508,6 +514,124 @@ def flow_from_dataframe(self, node_targets, shuffle=False):
...
+        batch_size (int): Size of batch to return.
+        name (str or None): Name of the generator (optional).
+    """
+
+    def __init__(self, G, batch_size, name=None):
+        if not isinstance(G, StellarGraphBase):
+            raise TypeError("Graph must be a StellarGraph object.")
+
+        self.graph = G
+        self.num_samples = [0]  # for compatibility with GraphSAGE
```
[the remark this answers is not in the document]
Comment (daokunzhang) [reply stellargraph/mapper/node_mappers.py]: The classes of LinkSequence, OnDemandLinkSequence and NodeSequence need the parameter num_samples in their init functions to set the _sampling_schema parameter for GraphSAGE. Actually the num_samples parameter is not necessary for Attri2VecLinkGenerator or Attri2VecNodeGenerator, but we need to make some changes in the LinkSequence, OnDemandLinkSequence and NodeSequence classes. Do you think it is OK?

**Comment by daokunzhang** [comment]
Hi @PantelisElinas ,

I have finished the changes as per your comments.

The following is some explanation:

1. I keep the name of the class Attri2VecNodeGenerator unchanged. We can change it to NodeGenerator or other names when it is generalised to other cases in the future.

2. After merging the latest develop branch, the attri2vec node classification demo on DBLP with multiple workers still performs worse than single worker. I use the single worker to run attri2vec embeddings&node classification demos on both citeseer and DBLP.

3. I delete the unnecessary num_samples and sampling_schema parameters from Attri2VecNodeGenerator and Attri2VecLinkGenerator.

4. I add a function 'get_index_for_nodes' to the StellarGraphBase class to access node index for the convenience of getting the ids of context node in attri2vec.

If you need more changes, please let me know.

Best regards!
Daokun

File: stellargraph/mapper/node_mappers.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (PantelisElinas) [reply stellargraph/mapper/node_mappers.py]: Maybe NodeFeatureGenerator would be more informative; I just swapped the words Node and Feature for your suggestion.

File: stellargraph/mapper/node_mappers.py [hunk 6]
```diff
@@ -505,6 +537,114 @@ def flow_from_dataframe(self, node_targets, shuffle=False):
...
+        G.check_graph_for_ml()
+
+    def sample_features(self, head_nodes):
+        """
+        Sample content features of the head nodes, and return these as a list of feature 
+        arrays for the attri2vec algorithm.
+
+        Args:
+            head_nodes: An iterable of head nodes to perform sampling on.
+            sampling_schema: The sampling schema for the model, for compatibility with GraphSAGE and HinSAGE.
```
[the remark this answers is not in the document]
Comment (daokunzhang) [reply stellargraph/mapper/node_mappers.py]: I have deleted the sampling schema in the docstring.

File: tests/layer/test_attri2vec.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    # Construction from generator
+    G = example_graph_1(feature_size=3)
+    gen = Attri2VecNodeGenerator(G, batch_size=2).flow([1, 2])
+    attri2vec = Attri2Vec(layer_sizes=[4, 8], generator=gen, bias=True)
+
+    assert attri2vec.dims == [3, 4, 8]
+    assert attri2vec.input_node_num == 4
+    assert attri2vec.n_layers == 2
+    assert attri2vec.bias
+
```
[the remark this answers is not in the document]
Comment (daokunzhang) [reply tests/layer/test_attri2vec.py]: Hi Pantelis,

If graphs have no features, the Attri2VecGenerator constructor will raise a Runtime Error, which has been tested in the test_attri2vec_nodemapper_constructor_no_feats() function in the test_node_mappers.py file.

**Comment by daokunzhang** [comment]
Hi @PantelisElinas,

Please check the latest update. I have fixed the typos and added the unit test for the get_index_for_nodes function.  I also deleted the attri2vec DBLP embedding learning and node classification demos.

If you need more changes, please let me know.

Best regards!
Daokun
